ethexe/compute/src/lib.rs

```diff
@@ -0,0 +1,299 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Sequencer for ethexe.
+
+use anyhow::{anyhow, Result};
+use ethexe_common::{
+    db::{BlockMetaStorage, CodesStorage},
+    events::{BlockEvent, BlockRequestEvent, RouterRequestEvent},
+    gear::{BlockCommitment, CodeCommitment, StateTransition},
+    BlockData,
+};
+use ethexe_db::{CodeInfo, Database};
+use ethexe_observer::Query;
+use ethexe_processor::{LocalOutcome, Processor};
+use futures::{future::BoxFuture, stream::FusedStream, FutureExt, Stream};
+use gprimitives::{CodeId, H256};
+use std::{
+    collections::VecDeque,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::task::JoinSet;
+
+#[derive(Debug)]
+pub struct BlockProcessed {
+    pub chain_head: H256,
+    // TODO (gsobol): remove commitments, this must be handled by validator if needed
+    pub commitments: Vec<BlockCommitment>,
+}
+
+#[derive(Debug)]
+pub enum ComputeEvent {
+    BlockProcessed(BlockProcessed),
+    CodeProcessed(CodeCommitment),
+}
+
+// TODO (gsobol): add state monitoring in prometheus
+// TODO (gsobol): append off-chain transactions handling
+// TODO (gsobol) asap: add tests for compute service
+pub struct ComputeService {
+    db: Database,
+    processor: Processor,
+    query: Query,
+    blocks_queue: VecDeque<BlockData>,
+    process_block: Option<BoxFuture<'static, Result<BlockProcessed>>>,
+    process_codes: JoinSet<Result<CodeCommitment>>,
+}
+
+impl Stream for ComputeService {
+    type Item = Result<ComputeEvent>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if let Some(Poll::Ready(res)) = self.process_block.as_mut().map(|f| f.poll_unpin(cx)) {
+            if let Some(block) = self.blocks_queue.pop_front() {
+                let context = ChainHeadProcessContext {
+                    db: self.db.clone(),
+                    processor: self.processor.clone(),
+                    query: self.query.clone(),
+                };
+
+                self.process_block = Some(Box::pin(context.process(block)));
+            } else {
+                self.process_block = None;
+            }
+
+            return Poll::Ready(Some(res.map(ComputeEvent::BlockProcessed)));
+        }
+
+        if let Poll::Ready(Some(res)) = self.process_codes.poll_join_next(cx) {
+            let res = match res {
+                Ok(res) => res.map(ComputeEvent::CodeProcessed),
+                Err(err) => Err(err.into()),
+            };
+            return Poll::Ready(Some(res));
+        }
+
+        Poll::Pending
+    }
+}
+
+impl FusedStream for ComputeService {
+    fn is_terminated(&self) -> bool {
+        false
+    }
+}
+
+impl ComputeService {
+    pub fn new(db: Database, processor: Processor, query: Query) -> Self {
```

Comment:
I'd say that this service should be responsible for runtime instantiation, so only it creates processor (not any external essense)

Reply:
I would better to add TODO in order to think about. I don't think compute should be used in rpc calls and processor should be used there directly.

Reply:
Nice idea about todo. But for production actions processor shouldn't be used in rpc. For case of read states etc there's no problem